crates/api/src/error.rs

```diff
@@ -0,0 +1,127 @@
+//! Kitsune2 error types.
+
+use std::sync::Arc;
+
+/// A clonable trait-object inner error.
+#[derive(Clone, Default)]
+pub struct DynInnerError(
+    pub Option<Arc<dyn std::error::Error + 'static + Send + Sync>>,
+);
+
+impl std::fmt::Debug for DynInnerError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl std::fmt::Display for DynInnerError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.0.as_ref() {
+            None => f.write_str("None"),
+            Some(s) => s.fmt(f),
+        }
+    }
+}
+
+impl std::error::Error for DynInnerError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        self.0.as_ref().map(|s| {
+            let out: &(dyn std::error::Error + 'static) = &**s;
+            out
+        })
+    }
+}
+
+impl DynInnerError {
+    /// Construct a new DynInnerError from a source error.
+    pub fn new<E: std::error::Error + 'static + Send + Sync>(e: E) -> Self {
+        Self(Some(Arc::new(e)))
+    }
+}
+
+/// The core kitsune2 error type. This type is used in all external
+/// kitsune apis as well as internally in some modules.
+///
+/// This type is required to implement `Clone` to ease the use of
+/// shared futures, which require the entire `Result` to be `Clone`.
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum K2Error {
+    /// Generic Kitsune2 internal error.
+    #[error("{ctx} (src: {src})")]
+    Other {
+        /// Any context associated with this error.
+        ctx: Arc<str>,
```

Comment:
Why not a `String`?

Reply:
Agree, we'll want to be able to format values in to error messages sometimes

Reply:
The ref pointer makes cloning cheaper, and Arc or Box str is a type indication that the string won't be change-able.

Take a look at the `other()` and `other_src()` helper functions which I imagine should be used most of the time for constructing. If you need to do it manually, you totally can.

All that being said, I'm fine switching back to Strings if we don't care about the clone overhead.

Reply:
Got it, we can do what we need to do then, sounds good!